adafruit_gc_iot_core.py

```diff
@@ -0,0 +1,390 @@
+        else:
+            raise TypeError("This library requires a NetworkManager object.")
+        # Validate Secrets
+        if hasattr(secrets, "keys"):
+            self._secrets = secrets
+        else:
+            raise AttributeError(
+                "Project settings are kept in secrets.py, please add them there!"
+            )
+        self._logger = None
+        if log is True:
+            self._logger = logging.getLogger("log")
+            self._logger.setLevel(logging.DEBUG)
+        # Configuration, from secrets file
+        self._proj_id = secrets["project_id"]
+        self._region = secrets["cloud_region"]
+        self._reg_id = secrets["registry_id"]
+        self._device_id = secrets["device_id"]
+        self._private_key = secrets["private_key"]
+        self.broker = "mqtt.googleapis.com"
+        self.username = b"unused"
+        self.cid = self.client_id
+
+    @property
+    def client_id(self):
+        """Returns a Google Cloud IOT Core Client ID.
+        """
+        client_id = "projects/{0}/locations/{1}/registries/{2}/devices/{3}".format(
+            self._proj_id, self._region, self._reg_id, self._device_id
+        )
+        if self._logger:
+            self._logger.debug("Client ID: {}".format(client_id))
+        return client_id
+
+
+    def generate_jwt(self, ttl=43200, algo="RS256"):
+        """Generates a JSON Web Token (https://jwt.io/) using network time.
+        :param int jwt_ttl: When the JWT token expires, defaults to 43200 minutes (or 12 hours).
+        :param str algo: Algorithm used to create a JSON Web Token.
+
+        Example usage of generating and setting a JSON-Web-Token:
+        ..code-block:: python
+
+            jwt = CloudCore.generate_jwt()
+            print("Generated JWT: ", jwt)
+        """
+        if self._logger:
+            self._logger.debug("Generating JWT...")
+        self._get_local_time()
+        claims = {
+            # The time that the token was issued at
+            "iat": time.time(),
+            # The time the token expires.
+            "exp": time.time() + ttl,
+            # The audience field should always be set to the GCP project id.
+            "aud": self._proj_id,
+        }
+        jwt = JWT.generate(claims, self._private_key, algo)
+        return jwt
+
+    # pylint: disable=line-too-long, too-many-locals
+    def _get_local_time(self):
+        """Fetch and "set" the local time of this microcontroller to the
+        local time at the location, using an internet time API.
+        from Adafruit IO Arduino
+        """
+        api_url = None
+        try:
+            aio_username = self._secrets["aio_username"]
+            aio_key = self._secrets["aio_key"]
+        except KeyError:
+            raise KeyError(
+                "\n\nOur time service requires a login/password to rate-limit. Please register for a free adafruit.io account and place the user/key in your secrets file under 'aio_username' and 'aio_key'"
```

Comment:
hmm id kinda prefer not do also depend on aio. do you need the true local time or just NTP time?

Reply:
@ladyada  We'd need the NTP time, google uses http://time.google.com in their examples.

Reply:
i think the ESP32 has a built in NTP function...check it?

Reply:
@ladyada  the Google IoT Micropython library calls NTPTime (https://github.com/micropython/micropython/blob/master/ports/esp8266/modules/ntptime.py#L29)  which grabs the NTP time and sets the device's local time (as UTC time).